packages/backend/src/services/ProjectService/ProjectService.ts

```diff
@@ -733,19 +734,79 @@ export class ProjectService {
         return { adapter, sshTunnel };
     }
 
+    static updateExploreWithGranularity(
+        explore: Explore,
+        metricQuery: MetricQuery,
+        warehouseClient: WarehouseClient,
+        granularity?: DateGranularity,
+    ): Explore {
+        if (granularity) {
+            const timeDimensionsMap: Record<string, CompiledDimension> =
+                Object.values(explore.tables).reduce<
+                    Record<string, CompiledDimension>
+                >((acc, t) => {
+                    Object.values(t.dimensions).forEach((dim) => {
+                        if (
+                            dim.type === DimensionType.TIMESTAMP ||
+                            dim.type === DimensionType.DATE
+                        ) {
+                            acc[getItemId(dim)] = dim;
+                        }
+                    });
+                    return acc;
+                }, {});
+
+            const firstTimeDimensionIdInMetricQuery =
+                metricQuery.dimensions.find(
+                    (dimension) => !!timeDimensionsMap[dimension],
+                );
+            if (firstTimeDimensionIdInMetricQuery) {
+                const dimToOverride =
+                    timeDimensionsMap[firstTimeDimensionIdInMetricQuery];
+                const { baseDimensionId } = getDateDimension(
+                    firstTimeDimensionIdInMetricQuery,
+                );
+                const baseTimeDimension =
+                    dimToOverride.timeInterval && baseDimensionId
+                        ? timeDimensionsMap[baseDimensionId]
+                        : dimToOverride;
+                const dimWithGranularityOverride =
+                    createDimensionWithGranularity(
+                        dimToOverride.name,
+                        baseTimeDimension,
+                        explore,
+                        warehouseClient,
+                        granularity,
+                    );
+                return replaceDimensionInExplore(
+                    explore,
+                    dimWithGranularityOverride,
+                );
+            }
+        }
+        return explore;
+    }
+
     private static async _compileQuery(
         metricQuery: MetricQuery,
         explore: Explore,
         warehouseClient: WarehouseClient,
         userAttributes: UserAttributeValueMap,
+        granularity?: DateGranularity,
     ): Promise<CompiledQuery> {
-        const compiledMetricQuery = compileMetricQuery({
+        const exploreWithOverride = ProjectService.updateExploreWithGranularity(
```

Comment:
Before compiling the SQL we update the explore to have the dimension with the new granularity